### PR TITLE
Please review my PR for fyle oauth flask template 

### PR DESCRIPTION
**Pull request by subhajit21**
Client Id ,Client Secret and base_url for FyleSDK connection are accepting value  by Environment variable .Others necessary URLs are hard coded . Hardcoded URLs can be removed and can accept value by Environment Variable . Displaying Employee data in basic HTML page .
Please review my PR.

### CODE REVIEW

File: FlaskApp/routes.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+from FlaskApp import app
+import json
+import requests
+
+
+
+# Necessary URLs and Client Id and Client Secret
+
+REDIRECT_URL="http://localhost:8080/profile"
+CLIENT_ID= os.environ.get("client_id")
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: keep all the environment variable names in Block letters
eg. os.environ.get('CLIENT_ID')

File: FlaskApp/routes.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import os
+from flask import Flask,render_template,redirect,url_for,request
+from fylesdk import FyleSDK
+from FlaskApp import app
+import json
+import requests
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: Keep all the internal library imports at the top and add a line separator

eg.
```
import os
import json
import requests

from fylesdk import FyleSDK

from FlaskApp import app
from flask import Flask, render_template, redirect, url_for, request (there no spaces after comma in code, please add that)

File: FlaskApp/routes.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+
+REDIRECT_URL = os.environ.get("REDIRECT_URL")
+CLIENT_ID = os.environ.get("CLIENT_ID")
+CLIENT_SECRET = os.environ.get("CLIENT_SECRET")
+BASE_URL = os.environ.get("BASE_URL")
+AUTHORIZE_URL = "{0}/app/main/#/oauth/authorize?".format(BASE_URL)
+APP_URL = "{0}client_id={1}&redirect_uri={2}&response_type=code&state=ajsfbjak".format(
+    AUTHORIZE_URL, CLIENT_ID, REDIRECT_URL)
+ENDPOINT = '{0}/api/oauth/token'.format(BASE_URL)
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: Rename to TOKEN_URL

File: FlaskApp/routes.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+# Necessary URLs and Client Id and Client Secret
+
+
+REDIRECT_URL = os.environ.get("REDIRECT_URL")
+CLIENT_ID = os.environ.get("CLIENT_ID")
+CLIENT_SECRET = os.environ.get("CLIENT_SECRET")
+BASE_URL = os.environ.get("BASE_URL")
+AUTHORIZE_URL = "{0}/app/main/#/oauth/authorize?".format(BASE_URL)
+APP_URL = "{0}client_id={1}&redirect_uri={2}&response_type=code&state=ajsfbjak".format(
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: No need to separate APP_URL
you can just do
```
"{0}/app/main/#/oauth/authorize?client_id={1}&redirect_uri={2}&response_type=code&state=ajsfbjak".format(BASE_URL, CLIENT_ID, REDIRECT_URI)

File: FlaskApp/routes.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import requests
+from flask import render_template, redirect, url_for, request
+from fylesdk import FyleSDK
+
+from FlaskApp import app
+
+# Necessary URLs and Client Id and Client Secret
+
+
+REDIRECT_URL = os.environ.get("REDIRECT_URL")
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: rename to `REDIRECT_URI`

File: FlaskApp/routes.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    AUTHORIZE_URL, CLIENT_ID, REDIRECT_URL)
+ENDPOINT = '{0}/api/oauth/token'.format(BASE_URL)
+
+
+# Home/index route, Only Accepts GET and POST method
+# For GET method it redirects home Page (welcome.html)
+# For POST method , it redirects APP_URL, with code 302
+
+@app.route('/', methods=['GET', 'POST'])
+@app.route('/home', methods=['GET', 'POST'])
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: rename to `/login`

File: FlaskApp/routes.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+'''Making connection with  FyleSDK
+   Fetching the profile details of Employee and returning necessary details  
+   base_url for FyleSDK connection is 'https://staging.fyle.in/' '''
+
+
+def get_profile_details(the_refrsh_token):
+    global CLIENT_ID
+    global CLIENT_SECRET
+    global BASE_URL
+    connection = FyleSDK(
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: can you add a separate file with class `Fyle` and in its constructor(`__init__()`) make the connection and use the reference

File: FlaskApp/routes.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    error = None
+    error = request.args.get('error')
+    code = request.args.get('code')
+    if request.method == 'GET' and code:
+        json_response = requests.post(ENDPOINT, data={"grant_type": "authorization_code",
+                                                      "client_id": CLIENT_ID,
+                                                      "client_secret": CLIENT_SECRET,
+                                                      "code": code})
+        data = json.loads(json_response.text)
+        my_refresh_token = data.get("refresh_token")
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: rename to refresh_token

File: FlaskApp/routes.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        return render_template('profile.html', emp_data=emp_details)
+    return redirect(url_for('login'))
+
+
+'''Making connection with  FyleSDK
+   Fetching the profile details of Employee and returning necessary details  
+   base_url for FyleSDK connection is 'https://staging.fyle.in/' '''
+
+
+def get_profile_details(the_refrsh_token):
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: rename to refresh_token

File: FlaskApp/routes.py (hunk 6 above)
Comment (Shwetabhk) [inline FlaskApp/routes.py]: don't use `POST` and `GET` in the same call, for post make a view with `/authorize` endpoints

File: FlaskApp/routes.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    global CLIENT_ID
+    global CLIENT_SECRET
+    global BASE_URL
+    connection = FyleSDK(
+        base_url=BASE_URL,
+        client_id=CLIENT_ID,
+        client_secret=CLIENT_SECRET,
+        refresh_token=the_refrsh_token
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: rename `refresh_token`

File: FlaskApp/routes.py (hunk 9 above)
Comment (Shwetabhk) [inline FlaskApp/routes.py]: move this function in the Fyle class in a separate file

File: FlaskApp/routes.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+from FlaskApp import fyle_connection
+
+# Necessary URLs and Client Id and Client Secret
+
+
+REDIRECT_URI = os.environ.get("REDIRECT_URL")
+CLIENT_ID = os.environ.get("CLIENT_ID")
+CLIENT_SECRET = os.environ.get("CLIENT_SECRET")
+BASE_URL = os.environ.get("BASE_URL")
+#AUTHORIZE_URL = "".format(BASE_URL)
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: you can remove this

File: FlaskApp/fyle_connection.py [hunk 12]
```diff
@@ -0,0 +1,22 @@
+import requests
```
Comment (Shwetabhk) [inline FlaskApp/fyle_connection.py]: put this file in a folder called `connectors` and file name `fyle.py`

File: FlaskApp/routes.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+AUTHORIZE_URL = "{0}/app/main/#/oauth/authorize?client_id={1}&redirect_uri={2}&response_type=code&state=ajsfbjak".format(
+    BASE_URL, CLIENT_ID, REDIRECT_URI)
+TOKEN_URL = '{0}/api/oauth/token'.format(BASE_URL)
+
+
+# Make authorization
+
+@app.route('/authorize', methods=['POST'])
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: copy this and move it below the home view

File: FlaskApp/routes.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+from fylesdk import FyleSDK
+
+from FlaskApp import app
+from FlaskApp import fyle_connection
+
+# Necessary URLs and Client Id and Client Secret
+
+
+REDIRECT_URI = os.environ.get("REDIRECT_URL")
+CLIENT_ID = os.environ.get("CLIENT_ID")
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: directly call the environment variable in the FyleConnector class

File: FlaskApp/routes.py [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+    code = request.args.get('code')
+    if code:
+        json_response = requests.post(TOKEN_URL, data={"grant_type": "authorization_code",
+                                                       "client_id": CLIENT_ID,
+                                                       "client_secret": CLIENT_SECRET,
+                                                       "code": code})
+        data = json.loads(json_response.text)
+        refresh_token = data.get("refresh_token")
+        fyle_connection = fyle_connection.FyleConnector(
+            BASE_URL, CLIENT_ID, CLIENT_SECRET, refresh_token)
```
Comment (Shwetabhk) [inline FlaskApp/routes.py]: no need to pass in the constructor, directly call the environment variables in the FyleConnector Class

File: FlaskApp/fyle_connection.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+        self.__refresh_token = refresh_token
+
+        self.__connection = FyleSDK(
+            base_url=self.__base_url,
+            client_id=self.__client_id,
+            client_secret=self.__client_secret,
+            refresh_token=self.__refresh_token
+        )
+
+    def emp_details(self):
```
Comment (Shwetabhk) [inline FlaskApp/fyle_connection.py]: rename to `get_employee_details`

File: FlaskApp/routes.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+import requests
+from flask import render_template, redirect, url_for, request
+from fylesdk import FyleSDK
+
+from FlaskApp import app
+from FlaskApp.connectors.fyle import FyleConnector
+
+# Necessary URLs and Client Id and Client Secret
+REDIRECT_URI = "http://localhost:8080/refresh_token"
```
Comment (snarayanank2) [inline FlaskApp/routes.py]: why is localhost hardcoded here?

File: FlaskApp/routes.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+def profile():
+    global REFRESH_TOKEN
+    if REFRESH_TOKEN:
+        fyle_connection = FyleConnector(REFRESH_TOKEN)
+        emp_details = fyle_connection.get_employee_details()
+        return render_template('profile.html', emp_data=emp_details)
+    return redirect(url_for('login'))
+
+# To get Refresh Token
+@app.route('/refresh_token', methods=['GET'])
```
Comment (snarayanank2) [inline FlaskApp/routes.py]: name it oauth_callback - that's a better url

File: FlaskApp/routes.py [hunk 19]
```diff
@@ -0,0 +1,85 @@
...
+    error = request.args.get('error')
+    code = request.args.get('code')
+    if code and error is None:
+        json_response = requests.post(TOKEN_URL, data={"grant_type": "authorization_code",
+                                                       "client_id": CLIENT_ID,
+                                                       "client_secret": CLIENT_SECRET,
+                                                       "code": code})
+        data = json.loads(json_response.text)
+        REFRESH_TOKEN = data.get("refresh_token")
+    return redirect(url_for('profile'))
```
Comment (snarayanank2) [inline FlaskApp/routes.py]: actually - I was wrong. this is actually a security hole.. if I login using this app, then the server saves my refresh token and then if you access the /profile page you'll actually see my profile information. don't save the refresh token, instead can you just render the profile right here. sorry about that.

File: FlaskApp/routes.py [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+
+
+'''Displaying Employee profile details 
+   If User allow to fetch Employee profile details We will get Code,error will be null
+   If User deny to fetch Employee profile details We will get an error(access denied ) , Code will be null
+   If User deny to fetch Employee profile details , He/She will be redirected to Home page(welcome.html) '''
+
+
+@app.route('/profile', methods=['GET'])
+def profile():
```
[the remark this answers is not in the document]
Comment (snarayanank2) [reply FlaskApp/routes.py]: this suggestion is incorrect. ignore.

**Pull request merged**